### ISSUE TITLE
Question mark in SQL query (postgres JSON operator) interpreted as prepared statement param

### ISSUE THREAD

**Issue by tonestrike** [metabase/metabase#1964]
I am trying to run this query: 

select \* from table where json_column->'some_key' ? 'some_value'

Every time I run any query containing ? on our postgres server, I get a "No value specified for parameter 1" error. We have run the exact same query in our database and it works perfectly.

⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by agilliland** [comment]
curious, i've never actually seen the question mark used as an operator in SQL before.  what SQL database are you using?

usually that operator is an indicator of a value that will be substituted in a prepared statement ... http://stackoverflow.com/questions/4773789/what-is-use-of-quesion-mark-in-sql

what exactly does your SQL statement above do?  is the `?` just checking that the key equals 'some_value'?

**Comment by tonestrike** [comment]
I'm not a postgres expert but here is a more realistic example of what we are trying to do: 

SELECT "t1"."id", "t1"."created_at", "t1"."updated_at", "t1"."enqueued_at", "t1"."dequeued_at", "t1"."expected_at", "t1"."schedule_at", "t1"."q_name", "t1"."data", "t1"."logs", "t1"."status", "t1"."meta" FROM "tasks" AS t1 WHERE ((("t1"."q_name" = 'action_log') AND ("t1"."data"->'action_ids' ? '7a99988007cb452b82374af8251ee7d7')) AND (("t1"."meta" -> 'user_id') = '3d3cdd3f6e195048b6bac9fd1e4f76cb'))

**Comment by camsaul** [comment]
`?` is used as the placeholder for parameters in prepared statements so it's assuiming that's what you mean. Since raw SQL queries aren't actually prepared we can probably just switch how execute them

**Comment by camsaul** [comment]
@agilliland [`?`](http://www.postgresql.org/docs/current/static/functions-json.html) is the Postgres JSONB operator to check whether an object contains a given key

**Comment by agilliland** [comment]
Yeah.  like you said Cam, we shouldn't be executing the native queries as prepared statements, that's probably the best way to fix this.

**Comment by agilliland** [comment]
After looking into this a little bit more it appears this is happening due to Postgres deviating from the SQL specification when using the `?` operator in their syntax.

http://stackoverflow.com/questions/14779896/does-the-jdbc-spec-prevent-from-being-used-as-an-operator-outside-of-quotes

It looks like you can simply use the `??` operator in your SQL instead and the PostgresJDBC driver will make the appropriate adaptations.  This isn't something that we are likely to address specifically in Metabase.

**Comment by camsaul** [comment]
We could still just execute raw SQL queries as non-prepared statements. That way whatever works in other tools like pgAdmin will work in Metabase.

**Comment by agilliland** [comment]
I agree, which is why I left this open.  

Given that this is caused by a database specific deviation from the SQL standard which is causing the problem and that there is a fairly simple workaround, I see this as a lower priority issue.  Happy to see someone pick it up and tackle it though.

**Comment by stas** [comment]
As a temporary solution we decided to use the array functions from postgres.
Here's an example:
```sql
SELECT id FROM column 
    WHERE ('SOME_ID' = ANY(SELECT jsonb_array_elements_text(data->'array_with_ids')))
limit 1 
;
```

Thank you for keeping this open!

**Comment by ghiculescu** [comment]
We've found the same issue using `hstore` type columns, and found that `??` works well where `?` doesn't.

`select * from table where hstore_column ? 'some_key'` <- fails, metabase thinks there's a missing parameter
`select * from table where hstore_column ?? 'some_key'` <- works!

**Comment by jeff303** [comment]
> We could still just execute raw SQL queries as non-prepared statements. That way whatever works in other tools like pgAdmin will work in Metabase.

It seems as though under some circumstances, even native queries can have parameters.  Unclear if this is exposed through the product or just a code path exercised by our tests, but it does complicate things somewhat.

**Comment by jeff303** [comment]
This should now be fixed (in 0.39) when executing native queries that don't have parameters, such as the original case outlined in this issue.
